File: leetcode/1883.go
func minSkips(dist []int, speed int, hoursBefore int) int {
	n := len(dist)
	dp := make([]int, n+1)
	dp[0] = 0
	for i := 1; i <= n; i++ {
		for j := i; j >= 0; j-- {
			temp := math.MaxInt64 / 2
			if j != i {
				temp = min(temp, ((dp[j]+dist[i-1]-1)/speed+1)*speed)
			}
			if j != 0 {
				temp = min(temp, dp[j-1]+dist[i-1])
			}
			dp[j] = temp
		}
	}
	for j := 0; j <= n; j++ {
		if dp[j] <= speed*hoursBefore {
			return j
		}
	}
	return -1
}
func min(x, y int) int {
	if x > y {
		return y
	}
	return x
}

File: leetcode/1883.cpp
class Solution {
private:
    using LL = long long;

public:
    int minSkips(vector<int>& dist, int speed, int hoursBefore) {
        int n = dist.size();
        vector<LL> f(n + 1, LLONG_MAX / 2);
        f[0] = 0;
        for (int i = 1; i <= n; ++i) {
            for (int j = i; j >= 0; --j) {
                LL temp = LLONG_MAX / 2;
                if (j != i) {
                    temp = min(temp, ((f[j] + dist[i - 1] - 1) / speed + 1) * speed);
                }
                if (j != 0) {
                    temp = min(temp, f[j - 1] + dist[i - 1]);
                }
                f[j] = temp;
            }
        }
        for (int j = 0; j <= n; ++j) {
            if (f[j] <= (LL)hoursBefore * speed) {
                return j;
            }
        }
        return -1;
    }
};
